### PR TITLE
Fix regression caused by #12084

### LINKED ISSUE #12327: AMP form issue
## What's the issue?

In last few hours AMP form doesn't send form variables within request.

## How do we reproduce the issue?

In second form, for example, if you fill data and press "Subscribe", entered data will not be sent to server and always same message will be shown: "Success! Thanks for trying the amp-form demo! Try to insert the word "error" as a name input in the form to see how amp-form handles errors.". Without {{name}} (in message) parameter, because this parameter was not sent to server. Same message is shown if you enter "error".

## What browsers are affected?

I tried Safari, Firefox and Chrome- all same.

## Which AMP version is affected?

amp-form-0.1.js

### PR DESCRIPTION
**Pull request by cvializ**
Fixes #12327

### CODE REVIEW

File: src/form-data-wrapper.js [hunk 1]
```diff
@@ -127,3 +127,13 @@ export class FormDataWrapper {
...
 }
+
+/**
+ * Check if the given object is a FormDataWrapper instance
+ * @param {*} o
+ * @return {boolean} True if the object is a FormDataWrapper instance.
+ */
+export function isFormDataWrapper(o) {
+  // instanceof doesn't work as expected, so we detect with duck-typing.
+  return !!o.getFormData;
```
Comment (choumx) [inline src/form-data-wrapper.js]: Minor: `typeof o.getFormData == 'function'`

File: src/form-data-wrapper.js [hunk 2]
```diff
@@ -127,3 +127,13 @@ export class FormDataWrapper {
     return this.formData_;
   }
 }
+
+/**
+ * Check if the given object is a FormDataWrapper instance
+ * @param {*} o
```
Comment (choumx) [inline src/form-data-wrapper.js]: Should this be `{!Object}`?

Comment (cvializ) [reply src/form-data-wrapper.js]: Probably, but Closure Compiler complains that `init.body` is `(Object|string|undefined)`, but I don't see why we'd make that the `@param` type. What do you think?

Comment (choumx) [reply src/form-data-wrapper.js]: This is fine with the truthy check you added. Still think that adding `typeof` is a bit more precise (and has precedence).

Comment (cvializ) [reply src/form-data-wrapper.js]: Coo, I just saw that. I like that better too

File: src/form-data-wrapper.js [hunk 3]
```diff
@@ -127,3 +127,13 @@ export class FormDataWrapper {
...
 }
+
+/**
+ * Check if the given object is a FormDataWrapper instance
+ * @param {!Object|string|undefined} o
+ * @return {boolean} True if the object is a FormDataWrapper instance.
+ */
+export function isFormDataWrapper(o) {
+  // instanceof doesn't work as expected, so we detect with duck-typing.
+  return !!o && typeof o.getFormData == 'function';
```
Comment (choumx) [inline src/form-data-wrapper.js]: Nit: No need for `!!` here. Also I meant `{*}` is fine but either way.

Comment (cvializ) [reply src/form-data-wrapper.js]: `check-types` complains that the type is incorrect without the `!!`. I'll keep it `*`

**Comment by zhangsu** [comment]
Thanks for the quick reaction! And very sorry about the breakage :(

Just curious: why doesn't `instanceof` work in this case? It looks like the `FormDataWrapper` imported in `xhr-impl.js` [F1] and the `FormDataWrapper` imported in `amp-form.js` [F2] are actually references to two different function objects, so the wrapper instance that `amp-form` sticks into `init.body` gets the prototype of F1, but the `instanceof` in `xhr-impl.js` tries to compare that prototype with the prototype of F2, which fails because `F1 !== F2`. Is this a transpiler issue?

**Comment by choumx** [comment]
Correct. It's because the file is bundled twice into different binaries: F1 in v0.js and F2 in amp-form.js. We've seen similar classes of bugs before and we're looking into why our existing checks didn't catch them and how else we can prevent it in the future.

**Pull request merged**